### PR TITLE
resethaltreq: Proposal for forcing a hart into debug mode out of reset

### PR DESCRIPTION
**Pull request by mwachs5**
Addresses #242 

Adds an optional bit in DMCONTROL analogous to `haltreq`, but is `resethaltreq`. This bit is not self-clearing, but only has impact when the hart's reset is deasserted.

### CODE REVIEW

**Comment by mwachs5** [comment]
I'm sketching out this implementation in rocket-chip. An immediate problem I see is that the debugger will have to make sure not accidentally clear this down as it's doing all its other operations, since it applies to all selected harts, but it's not clear how to read it back (the same problem exists for `haltreq` and `resumereq`, but those aren't usually long-lived bits, and they are specified as W-only).

**Comment by mwachs5** [comment]
I see in the discussion for #242 you mentioned having 2 separate "set" and "clear" bits. I guess that's why.

File: debug_module.tex [hunk 1]
```diff
@@ -140,6 +143,16 @@ \section{Run Control} \label{runcontrol}
 (How this is implemented is not further specified. A few
 clock cycles will be a more typical latency).
 
+The debugger can provide an optional \Fresethaltreq bit. Writing to this bit
```
Comment (timsifive) [inline debug_module.tex]: Debug Module, I assume

**Comment by mwachs5** [comment]
So I started to edit this to be set and clear but then I couldn't figure out how to make it optional. Are we ok with making this non-optional?

**Comment by timsifive** [comment]
I prefer to make it optional because if you have a lot of harts, routing all the signals could be a pain. Simplest way is to add a bit to dmstatus that, when 1, indicates the feature is supported.

**Comment by mwachs5** [comment]
@timsifive can I get an updated review? I think I did the intended set/clr changes.

File: debug_module.tex [hunk 2]
```diff
@@ -140,6 +143,19 @@ \section{Run Control} \label{runcontrol}
 (How this is implemented is not further specified. A few
 clock cycles will be a more typical latency).
 
+The debug module can implement optional halt-on-reset bits for each hart,
+which it indicates by setting \Fhasresethaltreq to 1.
+This means the debugger implements the \Fsetresethaltreq and \Fclrresethaltreq bits.
```
Comment (timsifive) [inline debug_module.tex]: DM, not debugger

File: debug_module.tex [hunk 3]
```diff
@@ -140,6 +143,19 @@ \section{Run Control} \label{runcontrol}
 (How this is implemented is not further specified. A few
 clock cycles will be a more typical latency).
 
+The debug module can implement optional halt-on-reset bits for each hart,
```
Comment (timsifive) [inline debug_module.tex]: Debug Module is capitalized

File: debug_module.tex [hunk 4]
```diff
@@ -140,6 +143,19 @@ \section{Run Control} \label{runcontrol}
...
+which it indicates by setting \Fhasresethaltreq to 1.
+This means the debugger implements the \Fsetresethaltreq and \Fclrresethaltreq bits.
+Writing 1 to \Fsetresethaltreq sets the halt-on-reset request bit for each
+selected hart.
+When a hart's halt-on-reset request bit is set, the hart will immediately enter
+debug mode on the next deassertion of its reset. This is true regardless of
+whether or not the hart's reset is caused by the debugger's
```
Comment (timsifive) [inline debug_module.tex]: DM's

**Pull request merged**